torchdata/nodes/io/file_read.py

```diff
@@ -0,0 +1,133 @@
+        ... )
+        >>>
+        >>> # Read binary files
+        >>> node = FileReader(source_node, mode="rb")
+        >>>
+        >>> # Read compressed files
+        >>> node = FileReader(
+        ...     source_node,
+        ...     transport_params={'compression': '.gz'}
+        ... )
+    """
+
+    SOURCE_KEY = "source"
+    DATA_KEY = "data"
+    METADATA_KEY = "metadata"
+
+    def __init__(
+        self,
+        source_node: BaseNode[Dict],
+        mode: str = "r",
+        encoding: Optional[str] = "utf-8",
+        transport_params: Optional[Dict] = None,
+    ):
+        """Initialize the FileReader.
+
+        Args:
+            source_node: Source node that yields dicts with file paths
+            mode: File open mode ('r' for text, 'rb' for binary)
+            encoding: Text encoding (None for binary mode)
+            transport_params: Parameters for smart_open transport layer
+                For S3:
+                    {'client': boto3.client('s3')}  # Use specific client
+                For compression:
+                    {'compression': '.gz'}  # Force gzip compression
+                    {'compression': '.bz2'}  # Force bz2 compression
+                    {'compression': 'disable'}  # Disable compression
+        """
+        super().__init__()
+        self.source = source_node
+        self.mode = mode
+        self.encoding = encoding
+        self.transport_params = transport_params or {}
+
+    def reset(self, initial_state: Optional[Dict[str, Any]] = None):
+        """Reset must fully initialize the node's state.
+
+        First resets source to beginning, then restores its state if provided.
+
+        Args:
+            initial_state: Optional state dictionary for resumption
+        """
+        super().reset(initial_state)
+        if initial_state is not None:
+            self.source.reset(initial_state.get(self.SOURCE_KEY))
+        else:
+            self.source.reset(None)
+
+    def next(self) -> Dict:
+        """Get next file path from source and return its contents."""
+        file_path_item = next(self.source)
+        file_path = file_path_item[self.DATA_KEY]
+
+        # Preserve metadata from source
+        source_metadata = file_path_item.get(self.METADATA_KEY, {})
+
+        try:
+            # Use smart_open to handle any file system
+            with open(
+                file_path,
+                self.mode,
+                encoding=None if "b" in self.mode else self.encoding,
+                transport_params=self.transport_params,
+            ) as f:
+                content = f.read()
+
+            # Create output with metadata
+            metadata = {"file_path": file_path}
+
+            # Include metadata from source
+            if source_metadata:
+                metadata.update(source_metadata)
+
+            return {self.DATA_KEY: content, self.METADATA_KEY: metadata}
+
+        except Exception as e:
```

Comment:
We can add a retry logic
what if there were some network hiccups, or some other retryable error
Is there a way to find out if the error was because of network (we can retry in that case) or some other factor (file not present, etc., so do not retry)